### ISSUE TITLE
Decide on a semantics of "support", "sample space", "extend by zero", and "restrict", and their implications for where/whether `logprob` should return `-Inf` or throw error

### ISSUE THREAD

**Issue by bzinberg** [probcomp/Gen.jl#206]
Continuation thread for the discussion started in https://github.com/probcomp/Gen/pull/203#discussion_r389749694.

**Comment by fsaad** [comment]
If the main character is "logprob" i.e., the behavior of the probability density function p : X → [0, ∞), then one standard terminology would be:
- domain: the set of values (X) on which the function p is defined;
- support: {x | p(x) > 0}.

"Sample space" usually refers to an abstract set Ω (e.g., the [0,1] or {0,1}^∞ is a canonical choice, endowed with the Lebesgue measure), on which every random variable Z : Ω → X is defined.  In Gen's case this space might be the sequence of bits generated by the (P)RNG.  There is rarely any interest in considering a sample space Ω other than [0,1], see e.g., Kallenberg Thm 3.19, Lemma 3.22, Thm 6.14, 6.16, and 6.17.

It seems logical at minimum to permit calling p(x) for any element x in the domain X, including those outside of the support.  Or you might chose to restrict the domain to the support, which also is sensible provided it is documented clearly, but results in a less general strange semantics of the CDF as being the the integral of the PDF for continuous distributions.

Whether calling p(x) for an x not in the domain, e.g., p("hello world") for p Gaussian or p(1.123) for p Categorical({"hello":0.4, "world":0.6}), results in "Error" or "0" is perhaps more debatable, since one can extend p to the universal set in the natural way if they have a good reason to do that.

**Comment by marcoct** [comment]
Copying and adapting from https://github.com/probcomp/Gen/pull/203

Each address is associated with a fixed measure space, which defines the 'base measure' for the random choices. We can decide to call the set of values in this space the "sample space" or the "domain" (maybe domain is more standard?). A `Distribution` type is associated with a base measure. The only non-trivial commitment here is to require that addresses are associated with fixed base measures, which is important.

It has also been helpful in Gen to have a notion of a "disciplined" generative function for which the **support** at an address is constant. But it is becoming more clear that is probably too restrictive to require this (or enforce it by throwing errors). However, I think that the notion of a disciplined generative function with constant support at addreses should be kept.

There is only a distinction between sample space and support when the support is not the "full" sample space. In the existing Gen usage and documentation, a "disciplined" modeling approach involves associating a fixed support with each address. That guarantees that 0 probability densities never show up. However, that makes some things inconvenient, which is why previously, 0 probability densities were sometimes allowed, and why, in this PR, they are allowed more consistently, instead of triggering an error.

I think it makes sense to admit a separate notion of sample space and support into Gen, where by "sample space" we mean the Julia type for the value of the random choice accepted by the `logpdf` Julia function, and by support we mean the subset of values of that type that have nonzero density. There are some meaningful differences in how they interact with the implementation -- e.g. the "sample space" is a static property of the distribution, and the support is not (it depends on the arguments).

Then, a "disciplined" generative function will associate addresses with fixed support. That is encouraged but not required. Gen will not enforce this by throwing errors when zero probability densities are encountered.

For reference, the relevant piece of existing documentation is [here](https://probcomp.github.io/Gen/dev/ref/modeling/#Choices-should-have-constant-support-1). It is a somewhat ambiguous about whether disciplined models are required or not (it uses the word "must" and then says that if you don't do it, some things might not work).

**Comment by marcoct** [comment]
This should also be documented under https://probcomp.github.io/Gen/dev/ref/distributions/

**Comment by marcoct** [comment]
At the end of the Distributions section, we can put a more technical treatment, where we discuss the base measure, measure space, densities, and support for distributions.

Also, there should eventually be documentation of the full measure-theoretic version of the GFI in the GFI section. We are working on a version of this that only requires elementary measure theory.

**Comment by marcoct** [comment]
Upon thinking more, it probably makes more sense to allow the sample space to be a subset of the Julia type expected by logpdf. That would allow additional user control over their "sample space", without having to define a complicated Julia type.

So then, there would be three sets of decreasing size (each is a subset of the previous):

- The set of values for which the Julia type matches. (This is a static property of the Distribution).

- The set of values for which logpdf does not raise an runtime error (this is the sample space / domain; which technically is defined by the distribution's measure space). (I think ideally this would be a static property of the distribution).

- The support. (This can be dynamic i.e. it can depend on the arguments)

**Comment by fsaad** [comment]
> The set of values for which logpdf does not raise an runtime error (this is the sample space / domain; which technically is defined by the distribution's measure space)

Since we are talking about "logpdf" then we should call it the "domain of the density function".

While we could technically talk about the "sample space" of a single random choice X with measure P_X by mentioning the probability triple (Ω, ℇ, P_X), I am not a fan of this approach and it will very likely confuse readers who will think of "sample space" in the sense of the canonical probability space on [0,1], not the sample space of the probability triple obtained through the pushforward measure of the random choice.

In addition, since Gen features multiple random choices with dependence between them, the way that they are formally modeled is as a collection of random variables that are all defined on the same sample space (i.e., e.g., the unit interval or the PRNG).  This formal idea breaks down if we start referring to different random choices as being defined on different sample spaces.

**Comment by marcoct** [comment]
> While we could technically talk about the "sample space" of a single random choice X with measure P_X by mentioning the probability triple (Ω, ℇ, P_X), I am not a fan of this approach and it will very likely confuse readers who will think of "sample space" in the sense of the canonical probability space on [0,1], not the sample space of the probability triple obtained through the pushforward measure of the random choice.

Gen is formalized using a probability triple for each address that defines the base measure. Then the base measure on choice maps is derived via product and disjoint union of these per-address base measures. The question is what to call the first element of the probability triple for the base measure space for a single address.

> In addition, since Gen features multiple random choices with dependence between them, the way that they are formally modeled is as a collection of random variables that are all defined on the same sample space (i.e., e.g., the unit interval or the PRNG). This formal idea breaks down if we start referring to different random choices as being defined on different sample spaces.

No, we formally model the distribution on choice maps as above, using a base measure on choice maps that is derived from a base measure on each address in a set of possible addresses. The unit interval does not appear, and I think that formalizing it with unit intervals obfuscates the structure of the distribution.

**Comment by fsaad** [comment]
> No, we formally model the distribution on choice maps as above, using a base measure on choice maps that is derived from a base measure on each address in a set of possible addresses. The unit interval does not appear, and I think that formalizing obfuscates the structure of the distribution.

For the distribution on choice maps that is fine, but for the definition of the random choices themselves (i.e., the random variables) they have to (formally) be functions whose domain is a single canonical space so that Gen has a coherent semantics that adheres to standard concepts in probability theory, and also so that Gen can leverage known statistical results concerning e.g., kernels, sample paths, and various forms of convergence.

**Comment by bzinberg** [comment]
> For the distribution on choice maps that is fine, but for the definition of the random choices themselves (i.e., the random variables) they have to (formally) be functions whose domain is a single canonical space

Why can't different random variates (avoiding the loaded term "random variable") have different sample spaces, and each trace have a different sample space, and we say that there must be a canonical isomorphism between the marginal distribution of a given choice within a given trace and the documented definition of the random variate?

**Comment by marcoct** [comment]
>  but for the definition of the random choices themselves (i.e., the random variables) they have to (formally) be functions whose domain is a single canonical space so that Gen has a coherent semantics that adheres to standard concepts in probability theory, and also so that Gen can leverage known statistical results concerning e.g., kernels, sample paths, and various forms of convergence.

I don't see why. Can you give an example? The standard results from probability theory would generally be applied to various measures and densities on choice maps, not to distributions on single random choices. In fact, I think Gen and its inference programs can generally be formalized without introducing a notion of "random variable" at all. Probability results can be applied to the measures and kernels on choice maps, without as much dependence on how that measure is structured as you are implying by invoking "random variables".

**Comment by fsaad** [comment]
> Gen is formalized using a probability triple for each address that defines the base measure.

It does not sound right to use the term "multiple probability triples"---at least in terms of usual terminology from measure theoretic probability, there is only one probability triple (equiv. one probability space) on which the whole random process is defined.

That is not to say that Gen cannot have its own terminology or conventions about this matter, provided the documentation notes that the usage is different from what a reader might otherwise expect.

**Comment by marcoct** [comment]
> It does not sound right to use the term "multiple probability triples"---at least in terms of usual terminology from measure theoretic probability, there is only one probability triple (equiv. one probability space) on which the whole random process is defined.

There is still just one probability triple, but its measurable space is constructed from simpler measurable spaces/

> That is not to say that Gen cannot have its own terminology or conventions about this matter, provided the documentation notes that the usage is different from what a reader might otherwise expect.

Agreed, and thanks for point out that this construction seems unusual to you. I think this discussion would be more efficient with a section of the documentation on the measure theoretic formalization of the GFI. This is something we have made progress on, but it's not quite ready yet.

**Comment by fsaad** [comment]
> I don't see why. Can you give an example? The standard results from probability theory would generally be applied to various measures and densities on choice maps, not to distributions on single random choices.

OK we can talk more about this offline.  My sense is that we need sequences of random variables on the same probability space if we want to make statements about limits inside a measure, i.e., "Pr[ \lim_{n \to \infty} X_n]" (which appear in a.s. convergence or Borel-Cantelli/0-1 laws), versus taking limits of measures i.e., "\lim_{n \to \infty} Pr[X_n]" (as in weak convergence).

It does seem important to pin this down especially since Gen is Turing complete, it can generate infinite sequences of random variables.

**Comment by fsaad** [comment]
> since Gen is Turing complete, it can generate infinite sequences of random variables.

Specifically, consider the following stochastic procedure for estimating a sample mean using a lazily generated sequence of i.i.d samples

```python
def sample_mean():
    n = 0; s = 0;
    while True:
        n += 1
        x = @trace(normal(0,1), :n => :x)
        s = s + x
        z = @trace(normal(s/n, 0.01), :n => :z)
        yield z
```
We would like to be able to make a statement about the almost-sure convergence of the sequence of random choices [(1 => :z), ( 2 => :z), (3 => :z), ...] to the mean 0 of the standard normal, or something close to the mean (since :n => :z is noisy around the running sum s).

I was specifically concerned about "pushing-forward" the Lebesgue on [0,1] through the measurable map from ([0,1], B([0,1])) to the infinite dimensional space (R^∞, B(^R∞)) that is defined by the program.  It's not clear to me how to do probabilistic programming with densities in this case, since [there is no](https://en.wikipedia.org/wiki/Infinite-dimensional_Lebesgue_measure) infinite-dimensional Lebesgue measure to use as the stock measure (the pushforward measure can be used to define the distribution on the sequences, but it's not clear that it has any useful structure without first knowing that the random variables (:n => :x) are measurable maps on the same sample space).

That said, this example is not really relevant to Gen, which does not have generators or the notion of a trace as a sample path of a stochastic process like the one above.

**Comment by bzinberg** [comment]
@fsaad, I would say a Gen program doesn't define a distribution on infinite sequences, it defines a distribution on a space Ω of traces, such that each trace in Ω is finite but there might not be an upper bound for the size of traces in Ω.  Even in classical models of computation, infinitary computation is very much its own thing (e.g. ordinary Turing machines don't do it).

**Comment by marcoct** [comment]
> It does seem important to pin this down especially since Gen is Turing complete, it can generate infinite sequences of random variables.

Right. But for a generative function to be valid, it has to halt with probability 1. That is, there needs to exist a probability measure on the space of traces (which are all finite).

> Specifically, consider the following stochastic procedure for estimating a sample mean using a lazily generated sequence of i.i.d samples

It's interesting that it's possible for a probabilistic programming language to ascribe meaningful semantics to infinite sequences and non-halting executions. I wonder if there's existing work in that area.

**Comment by fsaad** [comment]
@bzinberg @marcoct I agree.  In my example I used `yield` to capture the idea that the infinite computation is "frozen" as a generator,. and the trace can be lazily extended.

> such that each trace in Ω is finite but there might not be an upper bound for the size of traces in Ω

For Gen, I agree with you.  The implication of this definition is that we cannot make statements about almost sure convergence of sequences of traces (or their truncations):

 +   Pr[ For any ϵ, there exists n such that |Xm - u| < ϵ whenever m > n] = 1

but we can about weak convergence, by studying the truncations:

  +  For any ϵ, δ, there exists n such that Pr[ |Xm - u| < ϵ ] > 1 - δ whenever m > n.

**Comment by bzinberg** [comment]
Ah, got it.

> The implication of this definition is that we cannot make statements about almost sure convergence of sequences of traces (or their truncations)

I think we can still talk mathematically about the [direct limit](https://en.wikipedia.org/wiki/Direct_limit) of an infinite diagram of Ω's, each of which is a set of Gen traces with bounded running time.  Elements of that direct limit could perhaps be viewed as infinitary computations, even though Gen cannot represent them as traces.
